conda/core/package_cache_data.py

```diff
@@ -782,143 +785,156 @@ def execute(self):
 
-                future = fetch_executor.submit(
-                    do_cache_action,
-                    prec_or_spec,
+            def cancelled():
+                """
+                Used to cancel download threads.
+                """
+                nonlocal cancelled_flag
+                return cancelled_flag
+
+            with signal_handler(conda_signal_handler), time_recorder(
+                "fetch_extract_execute"
+            ), ThreadPoolExecutor(
+                context.fetch_threads
+            ) as fetch_executor, ThreadPoolExecutor(
+                EXTRACT_THREADS
+            ) as extract_executor:
+                for prec_or_spec, (
                     cache_action,
-                    progress_bar,
-                    cancelled=cancelled,
-                )
-
-                future.add_done_callback(
-                    partial(
-                        done_callback,
-                        actions=(cache_action,),
-                        exceptions=exceptions,
-                        progress_bar=progress_bar,
-                        finish=False,
+                    extract_action,
+                ) in self.paired_actions.items():
+                    if cache_action is None and extract_action is None:
+                        # Not sure when this is reached.
+                        continue
+
+                    progress_bar = self._progress_bar(
+                        prec_or_spec, context_manager=pbar_context, leave=False
                     )
-                )
-                futures.append(future)
 
-            try:
-                for completed_future in as_completed(futures):
-                    futures.remove(completed_future)
-                    prec_or_spec = completed_future.result()
+                    progress_bars[prec_or_spec] = progress_bar
 
-                    cache_action, extract_action = self.paired_actions[prec_or_spec]
-                    extract_future = extract_executor.submit(
-                        do_extract_action,
+                    future = fetch_executor.submit(
+                        do_cache_action,
                         prec_or_spec,
-                        extract_action,
-                        progress_bars[prec_or_spec],
+                        cache_action,
+                        progress_bar,
+                        cancelled=cancelled,
                     )
-                    extract_future.add_done_callback(
+
+                    future.add_done_callback(
                         partial(
                             done_callback,
-                            actions=(cache_action, extract_action),
+                            actions=(cache_action,),
                             exceptions=exceptions,
-                            progress_bar=progress_bars[prec_or_spec],
-                            finish=True,
+                            progress_bar=progress_bar,
+                            finish=False,
```

Comment:
That change of the default to `finish`, why is this?

Reply:
To be honest, I cannot remember why that was done! I can revert it back though as I do not see any downsides.